packages/components/copybutton/src/CopyButton.tsx

```diff
@@ -44,76 +75,106 @@ export interface CopyButtonProps extends CommonProps {
    * @default medium
    */
   size?: 'small' | 'medium';
-}
+} & (StringValue | PromiseValue);
 
 function _CopyButton(
-  props: ExpandProps<CopyButtonProps>,
-  ref: React.Ref<HTMLDivElement>,
-) {
-  const {
+  {
+    buttonProps,
     onCopy,
     value,
     className,
+    label,
     testId = 'cf-ui-copy-button',
     tooltipText = 'Copy to clipboard',
     tooltipCopiedText = 'Copied!',
     tooltipProps,
     isDisabled = false,
+    preload,
     size = 'medium',
     ...otherProps
-  } = props;
-  const styles = getStyles({ size });
-
+  }: ExpandProps<CopyButtonProps>,
+  ref: React.Ref<HTMLDivElement>,
+) {
+  const styles = getCopyButtonStyles({ size });
+  const [isLoading, setIsLoading] = useState(false);
   const [copied, setCopied] = useState(false);
   const button = useRef<HTMLButtonElement | null>(null);
-  const handleOnCopy = useCallback(
-    (e: string) => {
-      if (onCopy) {
-        onCopy(e);
+  const resolvedValue = useRef<string | Promise<string>>(
+    typeof value === 'function' ? value() : value,
```

Comment:
It's kinda tricky for me to logic about this one. 
Why do we need to call `value` when we're resolving it within the `preload` useEffect?
Do you think it makes sense to just preload all the time and not make it optional?

Reply:
Personally I was thinking about not calling the value here and only doing it with the effect for full control, but this behaviour replicates the API from the web app. I'm not sure which I prefer, what if it's a really expensive API call, would we want to trigger that just on render? The button might never be clicked 🤔